### ISSUE TITLE
Defining custom gravity model

### ISSUE THREAD

**Issue by Stephan801** [tudat-team/tudat#127]
Hi,

i would like to use tudat in combination with tudatpy for education in geoscience. Is it possible to implement a custom gravity model like egm2008 or soon egm2020?

I found an old but still active issue under [https://github.com/Tudat/tudat/issues/594](https://github.com/Tudat/tudat/issues/594) but that seems to be with an old version without python interface. Preferably I would like to use python to make it easier for my pupils to work with this software. 

If there have to be made some additional codeing I freely volunteer to code it. I would just need a rough sketch at which points there has to be made an addition.

Thanks,
Stephan

**Comment by DominicDirkx** [comment]
Hi Stephan,

Great to hear that you want to use Tudat for your education. The short answer to your question is: yes, Tudat has this functionality. In fact, we recently made some changes to this part of the code, but it looks like we:
* Messed up the function naming
* Did not document everything
The functions you may want to use are here (predefined_spherical_harmonic and from_file_spherical_harmonic)
https://github.com/tudat-team/tudatpy/blob/develop/tudatpy/kernel/expose_numerical_simulation/expose_environment_setup/expose_gravity_field_setup.cpp#L213-L229
But, it looks like the function names should be the other way around. We'll correct this soon

At present one of these is also documented here:
https://py.api.tudat.space/en/latest/gravity_field.html#tudatpy.numerical_simulation.environment_setup.gravity_field.from_file_spherical_harmonic
which loads one of the gravity fields here:
https://github.com/tudat-team/tudat-resources/tree/master/resource/gravity_models
The at present undocumented function (now named predefined_spherical_harmonic; should be named from_file_spherical_harmonic) allows you to load gravity field files in the same format.

In short:
* You can use the existing tudatpy functions now for your purposes
* But, the one you need is not yet properly documented
* But, the one you need will have a different name soon
When we change the function names, and add the documentation, I'll post it here. The required changes look fairly minimal, so I expect we can do this soon.
Note that, as standard practice, these changes will first be pushed to the develop version of Tudat (see here: https://docs.tudat.space/en/stable/_src_getting_started/installation.html)

If you can, I would love to hear more about how you plan to use Tudat in your education (here or at d.dirkx_at_tudelft.nl). It's always great to see others use the tools we've developed!

Best,

Dominic

**Comment by DominicDirkx** [comment]
We've updated tudatpy in this commit:
https://github.com/tudat-team/tudatpy/commit/57895612170454fb72cb8e877498100d9979ab59
and the API docs in this commit:
https://github.com/tudat-team/tudat-multidoc/commit/ec7603a83e694a30dc51f2777149ef3eae951f26
The /dev version of tudatpy should be updated by tomorrow, as should the API docs.

**Comment by DominicDirkx** [comment]
@Stephan801, this functionality has now also been merged into the master branch. The latest version of tudatpy comes with this functionality.

**Comment by Stephan801** [comment]
Hi Dominic,

thanks for your work, that is fantastic and works like a charm!

I plan to use this software for teaching the basic concepts of gravitation and global positioning systems and how one affects the other. I hope to motivate some to contemplate about studying STEM. Since I teach pupils and not students I want to prepare a set of examples they just have to run. After seeing that you already prepared a lot of examples I thought about using tudat with tudatpy as interface, which would allow in an easy way to not only show the examples but allow them to play around and get a feeling. 


Best,
Stephan

**Comment by DominicDirkx** [comment]
Hi Stephan,

Glad to hear that it all works well :) 

It's really cool that you want to use Tudat to explain things to high(?) school kids! I'd love to hear about your experiences in doing so. Thus far, we have only used Tudat in our university curriculum, but with the recent(ish) move to Python, I hope the code had become more accesible.

Best,

Dominic

**Comment by Stephan801** [comment]
Hi @DominicDirkx ,

I teach people as part of a vocational training in a geoscience related subject, so a feeling for STEM is already given. In most cases the just don't have a highschool degree  and therefore don't know the math behind all. Over the last years I observed, that most of the time there seems to be little motiviation to learn higher mathematics, but not because they are not able to learn. There seems to be little motivation because they can not see why mathematics are needed and what can be done with it. 

Especially when teaching positioning systems, which is an unbelievable interesting but at the same time quite complex subject, A lot of my pupils are shocked, when they see the equations behind it. I want to develop a more playfull approach. by first showing them examples of related applications and then step by step explain the needed (simplified) mathematics directly in relation to the example, e.g. what is a time dependent variable or why integration is needed. Right now I do not expect everything to be understood down to the smallest detail, but I want to arise some interest and show that going to a college or university is not impossible just because someone chose to not go after a high school degree in the first place.

I was searching for inspiration on github and stumbled across tudat / tudatpy and all the examples you implemented. In my opinion they allow an easy access for people with little existing knowledge of astrodynamic to a simulation tool with scientific accuracy. This is remarkably fantastic.  Thanks for making them accessible for free! 

Best 

Stephan

**Comment by DominicDirkx** [comment]
Hi Stephan,

That's really cool, good luck in using Tudat in your trainings, and if you have any advice for additional documentation, features, etc., we're always open to idea :) 

Best,

Dominic
